File: packages/protocol/src/transfer-messages.ts
/**
 * JSON codec for the transfer control messages (design §5.4). These travel as the plaintext
 * inside AES-GCM frames whose header `type` names the message; `block_data` is raw file bytes
 * and is not handled here. Decoders validate shape so a decrypted-but-malformed frame fails
 * loudly rather than corrupting state.
 */

import {
  FrameType,
  type Ack,
  type BlockHash,
  type BlockRecv,
  type Complete,
  type Control,
  type Fail,
  type FileEntry,
  type Manifest,
  type Nack,
  type ResumeFileState,
  type ResumeState,
  type TransferMsg,
} from './transfer.js';
import { utf8 } from './bytes.js';

const decoder = new TextDecoder();

export function encodeControl(msg: TransferMsg): Uint8Array {
  return utf8(JSON.stringify(msg));
}

export function decodeControl(payload: Uint8Array): TransferMsg {
  let obj: unknown;
  try {
    obj = JSON.parse(decoder.decode(payload));
  } catch {
    throw new Error('control frame: invalid JSON');
  }
  if (typeof obj !== 'object' || obj === null) throw new Error('control frame: not an object');
  const m = obj as Record<string, unknown>;
  switch (m.type) {
    case FrameType.Manifest:
      return asManifest(m);
    case FrameType.BlockHash:
      return asBlockHash(m);
    case FrameType.BlockRecv:
      return asBlockRecv(m);
    case FrameType.Ack:
      return asAck(m);
    case FrameType.Nack:
      return asNack(m);
    case FrameType.Control:
      return asControl(m);
    case FrameType.Complete:
      return asComplete(m);
    case FrameType.Done:
      return { type: FrameType.Done };
    case FrameType.Fail:
      return asFail(m);
    case FrameType.ResumeState:
      return asResumeState(m);
    default:
      throw new Error(`control frame: unexpected type ${String(m.type)}`);
  }
}

function num(m: Record<string, unknown>, k: string): number {
  const v = m[k];
  if (typeof v !== 'number' || !Number.isFinite(v))
    throw new Error(`control frame: ${k} not a number`);
  return v;
}
function str(m: Record<string, unknown>, k: string): string {
  const v = m[k];
  if (typeof v !== 'string') throw new Error(`control frame: ${k} not a string`);
  return v;
}

function asFileEntry(v: unknown): FileEntry {
  if (typeof v !== 'object' || v === null) throw new Error('control frame: bad file entry');
  const f = v as Record<string, unknown>;
  return {
    idx: num(f, 'idx'),
    name: str(f, 'name'),
    size: num(f, 'size'),
    mime: str(f, 'mime'),
    lastModified: num(f, 'lastModified'),
    blockSize: num(f, 'blockSize'),
    blocks: num(f, 'blocks'),
    fileDigest: str(f, 'fileDigest'),
  };
}

function asManifest(m: Record<string, unknown>): Manifest {
  if (!Array.isArray(m.files) || m.files.length === 0)
    throw new Error('control frame: manifest.files');
  // transferId is optional; when present it must be a string and keeps its position right after
  // `type` so the re-encoded wire bytes match the original ordering.
  const transferId = m.transferId === undefined ? undefined : str(m, 'transferId');
  return {
    type: FrameType.Manifest,
    ...(transferId !== undefined ? { transferId } : {}),
    files: m.files.map(asFileEntry),
    totalSize: num(m, 'totalSize'),
  };
}
function asBlockHash(m: Record<string, unknown>): BlockHash {
  return {
    type: FrameType.BlockHash,
    fileIdx: num(m, 'fileIdx'),
    blockIdx: num(m, 'blockIdx'),
    sha256: str(m, 'sha256'),
  };
}
function asBlockRecv(m: Record<string, unknown>): BlockRecv {
  return { type: FrameType.BlockRecv, fileIdx: num(m, 'fileIdx'), blockIdx: num(m, 'blockIdx') };
}
function asAck(m: Record<string, unknown>): Ack {
  return { type: FrameType.Ack, fileIdx: num(m, 'fileIdx'), blockIdx: num(m, 'blockIdx') };
}
function asNack(m: Record<string, unknown>): Nack {
  const reason = str(m, 'reason');
  if (reason !== 'missing' && reason !== 'timeout') {
    throw new Error(`control frame: bad nack reason ${reason}`);
  }
  return {
    type: FrameType.Nack,
    fileIdx: num(m, 'fileIdx'),
    blockIdx: num(m, 'blockIdx'),
    reason,
  };
}
function asControl(m: Record<string, unknown>): Control {
  const op = str(m, 'op');
  if (op !== 'pause' && op !== 'resume' && op !== 'cancel') {
    throw new Error(`control frame: bad control op ${op}`);
  }
  return { type: FrameType.Control, op };
}
function asComplete(m: Record<string, unknown>): Complete {
  return { type: FrameType.Complete, fileDigest: str(m, 'fileDigest') };
}
function asResumeFileState(v: unknown): ResumeFileState {
  if (typeof v !== 'object' || v === null) throw new Error('control frame: bad resume file state');
  const f = v as Record<string, unknown>;
  return { idx: num(f, 'idx'), haveBlocks: num(f, 'haveBlocks') };
}
function asResumeState(m: Record<string, unknown>): ResumeState {
  if (!Array.isArray(m.files) || m.files.length === 0)
    throw new Error('control frame: resume_state.files');
  return {
    type: FrameType.ResumeState,
    transferId: str(m, 'transferId'),
    files: m.files.map(asResumeFileState),
  };
}
function asFail(m: Record<string, unknown>): Fail {
  const reason = str(m, 'reason');
  const allowed = [
    'digest_mismatch',
    'integrity',
    'sink_error',
    'canceled',
    'quota',
    'retry_exhausted',
  ];
  if (!allowed.includes(reason)) throw new Error(`control frame: bad fail reason ${reason}`);
  return { type: FrameType.Fail, reason: reason as Fail['reason'] };
}

File: packages/wire/transfer_messages.go
package wire

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
)

// JSON codec for the transfer control messages (design §5.4). These travel as the plaintext
// inside AES-GCM frames whose header Type names the message; block_data is raw file bytes and
// is not handled here. It is the Go twin of packages/protocol/src/transfer-messages.ts and
// must be byte-identical on the wire: field declaration order matches the TS object-literal
// order (so JSON keys serialize in the same sequence), and HTML escaping is disabled so
// characters like <, >, and & in a filename encode the same as JavaScript's JSON.stringify.

// ControlMsg is one decoded transfer control message. Callers type-switch on the concrete
// type; FrameType reports the frame-header tag to stamp when sending it.
type ControlMsg interface {
	FrameType() uint8
}

// FileEntry is one file's metadata within a manifest.
type FileEntry struct {
	Idx          int    `json:"idx"`
	Name         string `json:"name"` // sanitized on receipt
	Size         int64  `json:"size"`
	Mime         string `json:"mime"`
	LastModified int64  `json:"lastModified"`
	BlockSize    int    `json:"blockSize"`
	Blocks       int    `json:"blocks"`
	FileDigest   string `json:"fileDigest"` // canonical whole-file SHA-256 (hex)
}

// Manifest lists the files a transfer will deliver.
type Manifest struct {
	Type      uint8       `json:"type"`
	Files     []FileEntry `json:"files"`
	TotalSize int64       `json:"totalSize"`
}

// BlockHash carries a block's SHA-256 so the receiver can verify before acking.
type BlockHash struct {
	Type     uint8  `json:"type"`
	FileIdx  int    `json:"fileIdx"`
	BlockIdx int    `json:"blockIdx"`
	SHA256   string `json:"sha256"`
}

// BlockRecv is the window signal: the receiver has taken delivery of a block (pre-ack).
type BlockRecv struct {
	Type     uint8 `json:"type"`
	FileIdx  int   `json:"fileIdx"`
	BlockIdx int   `json:"blockIdx"`
}

// Complete tells the receiver every block was sent and gives the canonical digest to verify.
type Complete struct {
	Type       uint8  `json:"type"`
	FileDigest string `json:"fileDigest"`
}

// Done is the receiver's confirmation that verification succeeded.
type Done struct {
	Type uint8 `json:"type"`
}

// Fail aborts the transfer with a machine-readable reason.
type Fail struct {
	Type   uint8      `json:"type"`
	Reason FailReason `json:"reason"`
}

// FrameType returns the frame-header tag for each message; the JSON "type" field carries the
// same value, so the field is the single source of truth and cannot drift from the header.
func (m Manifest) FrameType() uint8 { return m.Type }

// FrameType reports the block_hash frame tag.
func (m BlockHash) FrameType() uint8 { return m.Type }

// FrameType reports the block_recv frame tag.
func (m BlockRecv) FrameType() uint8 { return m.Type }

// FrameType reports the complete frame tag.
func (m Complete) FrameType() uint8 { return m.Type }

// FrameType reports the done frame tag.
func (m Done) FrameType() uint8 { return m.Type }

// FrameType reports the fail frame tag.
func (m Fail) FrameType() uint8 { return m.Type }

// NewManifest builds a manifest message with the correct frame tag.
func NewManifest(files []FileEntry, totalSize int64) *Manifest {
	return &Manifest{Type: FrameManifest, Files: files, TotalSize: totalSize}
}

// NewBlockHash builds a block_hash message.
func NewBlockHash(fileIdx, blockIdx int, sha256 string) *BlockHash {
	return &BlockHash{Type: FrameBlockHash, FileIdx: fileIdx, BlockIdx: blockIdx, SHA256: sha256}
}

// NewBlockRecv builds a block_recv window signal.
func NewBlockRecv(fileIdx, blockIdx int) *BlockRecv {
	return &BlockRecv{Type: FrameBlockRecv, FileIdx: fileIdx, BlockIdx: blockIdx}
}

// NewComplete builds a complete message carrying the canonical file digest.
func NewComplete(fileDigest string) *Complete {
	return &Complete{Type: FrameComplete, FileDigest: fileDigest}
}

// NewDone builds a done message.
func NewDone() *Done { return &Done{Type: FrameDone} }

// NewFail builds a fail message with the given reason.
func NewFail(reason FailReason) *Fail { return &Fail{Type: FrameFail, Reason: reason} }

// EncodeControl serializes a control message to its wire JSON. HTML escaping is disabled and
// the encoder's trailing newline stripped so the bytes match JavaScript's JSON.stringify.
func EncodeControl(msg ControlMsg) ([]byte, error) {
	var buf bytes.Buffer
	enc := json.NewEncoder(&buf)
	enc.SetEscapeHTML(false)
	if err := enc.Encode(msg); err != nil {
		return nil, err
	}
	return bytes.TrimSuffix(buf.Bytes(), []byte("\n")), nil
}

// DecodeControl parses a control-frame payload, validating shape so a decrypted-but-malformed
// frame fails loudly rather than corrupting state. block_data is never passed here.
func DecodeControl(payload []byte) (ControlMsg, error) {
	var head struct {
		Type *int `json:"type"`
	}
	if err := json.Unmarshal(payload, &head); err != nil {
		return nil, errors.New("control frame: invalid JSON")
	}
	if head.Type == nil {
		return nil, errors.New("control frame: missing type")
	}
	switch uint8(*head.Type) {
	case FrameManifest:
		return decodeManifest(payload)
	case FrameBlockHash:
		return decodeBlockHash(payload)
	case FrameBlockRecv:
		return decodeBlockRecv(payload)
	case FrameComplete:
		return decodeComplete(payload)
	case FrameDone:
		return &Done{Type: FrameDone}, nil
	case FrameFail:
		return decodeFail(payload)
	default:
		return nil, fmt.Errorf("control frame: unexpected type %d", *head.Type)
	}
}

// rawFileEntry mirrors FileEntry with pointer fields so a missing key is distinguishable from
// a legitimate zero value (e.g. idx 0), matching the TS num()/str() presence checks.
type rawFileEntry struct {
	Idx          *int    `json:"idx"`
	Name         *string `json:"name"`
	Size         *int64  `json:"size"`
	Mime         *string `json:"mime"`
	LastModified *int64  `json:"lastModified"`
	BlockSize    *int    `json:"blockSize"`
	Blocks       *int    `json:"blocks"`
	FileDigest   *string `json:"fileDigest"`
}

func (r rawFileEntry) build() (FileEntry, error) {
	if r.Idx == nil || r.Name == nil || r.Size == nil || r.Mime == nil ||
		r.LastModified == nil || r.BlockSize == nil || r.Blocks == nil || r.FileDigest == nil {
		return FileEntry{}, errors.New("control frame: incomplete file entry")
	}
	return FileEntry{
		Idx: *r.Idx, Name: *r.Name, Size: *r.Size, Mime: *r.Mime,
		LastModified: *r.LastModified, BlockSize: *r.BlockSize, Blocks: *r.Blocks,
		FileDigest: *r.FileDigest,
	}, nil
}

func decodeManifest(payload []byte) (ControlMsg, error) {
	var raw struct {
		Files     []rawFileEntry `json:"files"`
		TotalSize *int64         `json:"totalSize"`
	}
	if err := json.Unmarshal(payload, &raw); err != nil {
		return nil, errors.New("control frame: invalid manifest")
	}
	if len(raw.Files) == 0 {
		return nil, errors.New("control frame: manifest.files empty")
	}
	if raw.TotalSize == nil {
		return nil, errors.New("control frame: manifest.totalSize missing")
	}
	files := make([]FileEntry, len(raw.Files))
	for i, rf := range raw.Files {
		f, err := rf.build()
		if err != nil {
			return nil, err
		}
		files[i] = f
	}
	return &Manifest{Type: FrameManifest, Files: files, TotalSize: *raw.TotalSize}, nil
}

func decodeBlockHash(payload []byte) (ControlMsg, error) {
	var raw struct {
		FileIdx  *int    `json:"fileIdx"`
		BlockIdx *int    `json:"blockIdx"`
		SHA256   *string `json:"sha256"`
	}
	if err := json.Unmarshal(payload, &raw); err != nil {
		return nil, errors.New("control frame: invalid block_hash")
	}
	if raw.FileIdx == nil || raw.BlockIdx == nil || raw.SHA256 == nil {
		return nil, errors.New("control frame: incomplete block_hash")
	}
	return &BlockHash{Type: FrameBlockHash, FileIdx: *raw.FileIdx, BlockIdx: *raw.BlockIdx, SHA256: *raw.SHA256}, nil
}

func decodeBlockRecv(payload []byte) (ControlMsg, error) {
	var raw struct {
		FileIdx  *int `json:"fileIdx"`
		BlockIdx *int `json:"blockIdx"`
	}
	if err := json.Unmarshal(payload, &raw); err != nil {
		return nil, errors.New("control frame: invalid block_recv")
	}
	if raw.FileIdx == nil || raw.BlockIdx == nil {
		return nil, errors.New("control frame: incomplete block_recv")
	}
	return &BlockRecv{Type: FrameBlockRecv, FileIdx: *raw.FileIdx, BlockIdx: *raw.BlockIdx}, nil
}

func decodeComplete(payload []byte) (ControlMsg, error) {
	var raw struct {
		FileDigest *string `json:"fileDigest"`
	}
	if err := json.Unmarshal(payload, &raw); err != nil {
		return nil, errors.New("control frame: invalid complete")
	}
	if raw.FileDigest == nil {
		return nil, errors.New("control frame: complete.fileDigest missing")
	}
	return &Complete{Type: FrameComplete, FileDigest: *raw.FileDigest}, nil
}

func decodeFail(payload []byte) (ControlMsg, error) {
	var raw struct {
		Reason *string `json:"reason"`
	}
	if err := json.Unmarshal(payload, &raw); err != nil {
		return nil, errors.New("control frame: invalid fail")
	}
	if raw.Reason == nil {
		return nil, errors.New("control frame: fail.reason missing")
	}
	switch FailReason(*raw.Reason) {
	case FailDigestMismatch, FailIntegrity, FailSinkError, FailCanceled, FailQuota:
		return &Fail{Type: FrameFail, Reason: FailReason(*raw.Reason)}, nil
	default:
		return nil, fmt.Errorf("control frame: bad fail reason %q", *raw.Reason)
	}
}
